### PR TITLE
Add GA4 tracking for content navigation

### PR DESCRIPTION
**Pull request by AshGDS**
Would you be able to approve this PR if all is OK? Thanks :+1:

## What

Adds the `ga4_tracking` option to three components. When this is set to true, it will add the relevant GA4 attributes.

## Why

https://trello.com/c/C9M2EJQu/553-content-navigation-chapter-headings-next-previous-view-a-printable-version-navigation-and-show-hide-updates-interaction

I have a branch in `government-frontend` that will use these attributes.

## Visual Changes

None.

### CODE REVIEW

File: app/views/govuk_publishing_components/components/_contents_list.html.erb [hunk 1]
```diff
@@ -12,10 +12,22 @@
...
+      event_name: "navigation",
+      type: "content",
+      section: t("components.contents_list.contents", locale: :en) || ""
+    }.to_json
+  end
   local_assigns[:aria] ||= {}
   component_helper = GovukPublishingComponents::Presenters::ComponentWrapperHelper.new(local_assigns)
   component_helper.add_class("gem-c-contents-list #{brand_helper.brand_class}")
-  component_helper.add_data_attribute({ module: "gem-track-click" })
+  component_helper.add_data_attribute({ module: modules })
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_contents_list.html.erb:29]: I _think_ that the code could be slightly reduced as follows (but I haven't worked on the wrapper for while so obviously check it works!).

```Ruby
ga4_tracking ||= false
if ga4_tracking
    ga4_data = {
      event_name: "navigation",
      type: "content",
      section: t("components.contents_list.contents", locale: :en) || ""
    }.to_json
end

...

component_helper.add_data_attribute({ module: "gem-track-click" })
component_helper.add_data_attribute({ module: "ga4-link-tracker" }) if ga4_tracking
component_helper.add_data_attribute({ ga4_link: ga4_data, ga4_track_links_only: "" }) if ga4_tracking
```

File: app/views/govuk_publishing_components/components/_contents_list.html.erb [hunk 2]
```diff
@@ -41,7 +53,8 @@
               track_label: contents_item[:href],
               track_options: {
                 dimension29: contents_item[:text]
-              }
+              },
+              ga4_link: {"index": {"index_link": position}, "index_total": contents.length}.to_json
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_contents_list.html.erb:57]: Does this also need some kind of `if ga4_tracking` check?

File: app/views/govuk_publishing_components/components/_previous_and_next_navigation.html.erb [hunk 3]
```diff
@@ -2,12 +2,16 @@
   add_gem_component_stylesheet("previous-and-next-navigation")
 
   if local_assigns.include?(:next_page) || local_assigns.include?(:previous_page)
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_previous_and_next_navigation.html.erb:4]: Unrelated: seems slightly strange to have an `if` statement begin here as part of a larger code block, was initially a bit stumped trying to understand this - I'd expect it to be a single erb statement i.e. `<% if ... %>`

File: app/views/govuk_publishing_components/components/_previous_and_next_navigation.html.erb [hunk 4]
```diff
@@ -2,12 +2,16 @@
...
+    ga4_tracking ||= false
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_previous_and_next_navigation.html.erb:5]: I'd move this above line 4 as I don't think it should be part of the `if` statement.

File: app/views/govuk_publishing_components/components/_previous_and_next_navigation.html.erb [hunk 5]
```diff
@@ -2,12 +2,16 @@
...
 %>
-  
+
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_previous_and_next_navigation.html.erb:7]: Nitpick: I'd remove this blank line entirely.

File: app/views/govuk_publishing_components/components/_previous_and_next_navigation.html.erb [hunk 6]
```diff
@@ -24,6 +28,14 @@
           data-track-label="<%= previous_page[:url] %>"
           data-track-dimension="previous"
           data-track-dimension-index="29"
+          <% if ga4_tracking %>
+            data-ga4-link = "<%= {
+              event_name: "navigation",
+              type: "content",
+              text: previous_page[:label],
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_previous_and_next_navigation.html.erb]: For both the next and previous labels, there's a check below to see if they exist. Can you write a quick test to check that this doesn't break if `ga4_tracking` is true but those labels don't exist?

File: app/views/govuk_publishing_components/components/_metadata.html.erb [hunk 7]
```diff
@@ -18,6 +18,8 @@
   classes << "direction-#{direction}" if local_assigns.include?(:direction)
   classes << "gem-c-metadata--inverse" if inverse
   classes << shared_helper.get_margin_bottom if local_assigns[:margin_bottom]
+
+  ga4_tracking ||= false
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_metadata.html.erb:22]: Going through this PR commit by commit so apologies if it comes later, but does this change require a test? Can it be included in this commit?

File: app/views/govuk_publishing_components/components/_previous_and_next_navigation.html.erb [hunk 8]
```diff
@@ -32,7 +32,7 @@
             data-ga4-link = "<%= {
               event_name: "navigation",
               type: "content",
-              text: previous_page[:label],
+              text: previous_page[:label] || title,
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_previous_and_next_navigation.html.erb:35]: Ah great, ignore my (previous but one?) comment.

**Review by andysellick — COMMENTED** [review]
All looks good, some minor comments/suggestions, can you also merge some of the commits together? e.g. yml + test + change in one commit, per component.

File: app/views/govuk_publishing_components/components/_contents_list.html.erb (hunk 2 above)
Comment (AshGDS) [reply app/views/govuk_publishing_components/components/_contents_list.html.erb:57]: Good spot, will add one.

File: app/views/govuk_publishing_components/components/_contents_list.html.erb (hunk 1 above)
Comment (AshGDS) [reply app/views/govuk_publishing_components/components/_contents_list.html.erb:29]: Thanks. I seem to need to keep the `ga4_data = nil` though, as I can't call `if ga4_tracking` within the `data` hash. It makes the template crash. Ruby doesn't render the HTML attribute if the value is `nil` so having the `ga4_data = nil` default seems to work.

File: app/views/govuk_publishing_components/components/_previous_and_next_navigation.html.erb (hunk 6 above)
Comment (AshGDS) [reply app/views/govuk_publishing_components/components/_previous_and_next_navigation.html.erb]: I added `|| title` in a future commit. Without the `|| title`, it was returning an empty string on the component renders that didn't have the label. Is this OK?

File: app/views/govuk_publishing_components/components/_previous_and_next_navigation.html.erb (hunk 8 above)
Comment (AshGDS) [reply app/views/govuk_publishing_components/components/_previous_and_next_navigation.html.erb:35]: Saw this after I replied to the other comment!

File: app/views/govuk_publishing_components/components/_metadata.html.erb (hunk 7 above)
Comment (AshGDS) [reply app/views/govuk_publishing_components/components/_metadata.html.erb:22]: Yep, it has a test :+1:

**Comment by AshGDS** [comment]
Thanks @andysellick I've made the changes - will squash everything once approved :+1:

File: app/views/govuk_publishing_components/components/_contents_list.html.erb [hunk 9]
```diff
@@ -12,10 +12,21 @@
...
+      type: "content",
+      section: t("components.contents_list.contents", locale: :en) || ""
+    }.to_json
+  end
   local_assigns[:aria] ||= {}
   component_helper = GovukPublishingComponents::Presenters::ComponentWrapperHelper.new(local_assigns)
   component_helper.add_class("gem-c-contents-list #{brand_helper.brand_class}")
   component_helper.add_data_attribute({ module: "gem-track-click" })
+  component_helper.add_data_attribute({ module: "ga4-link-tracker" }) if ga4_tracking
+  component_helper.add_data_attribute({ga4_link: ga4_data, ga4_track_links_only: ""}) if ga4_tracking
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_contents_list.html.erb:29]: Nitpick: can you add spaces to match the previous lines? i.e.

```
component_helper.add_data_attribute({ ga4_link: ga4_data, ga4_track_links_only: "" }) if ga4_tracking
```

File: app/views/govuk_publishing_components/components/_contents_list.html.erb [hunk 10]
```diff
@@ -32,7 +43,12 @@
     <ol class="gem-c-contents-list__list">
       <% contents.each.with_index(1) do |contents_item, position| %>
         <li class="<%= cl_helper.list_item_classes(contents_item, false) %>" <%= "aria-current=true" if contents_item[:active] %>>
-          <% link_text = format_numbers ? cl_helper.wrap_numbers_with_spans(contents_item[:text]) : contents_item[:text] %>
+          <% link_text = format_numbers ? cl_helper.wrap_numbers_with_spans(contents_item[:text]) : contents_item[:text]
+             ga4_link_data = nil
+             if ga4_tracking
+                ga4_link_data = {"index": {"index_link": position}, "index_total": contents.length}.to_json
+             end
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_contents_list.html.erb:50]: Should be able to shorten this? Also (nitpick) please add spaces to improve readability, as shown.

```Ruby
ga4_link_data = { "index": { "index_link": position }, "index_total": contents.length }.to_json if ga4_tracking
```

**Review by andysellick — APPROVED** [review]
Just a couple of small comments but happy to approve and let you sort them out and squash commits etc. 👍

**Comment by AshGDS** [comment]
Thanks @andysellick, squashed and fixed those two issues. Would you be able to approve the Percy change? Thanks!

**Pull request merged**